src/click/_compat.py

```diff
@@ -363,7 +363,7 @@ def get_text_stderr(
 
 
 def _wrap_io_open(
-    file: t.Union[str, "os.PathLike[str]", int],
+    file: t.Union[t.AnyStr, "os.PathLike[t.AnyStr]", int],
```

Comment:
My reasoning is in [this commit message](https://github.com/pallets/click/pull/2487/commits/29819e70fdcfbe457aab2da23dd2882e5ea76ad7).

The issue as I see it is that bytes already worked except when `atomic=True`. The change I made was to fix the types (removing `Any` etc) and then fix the implementation based on that. In particular, note *how few* non-type changes there are in that commit.

If you want to accept `str` only (why?), then the implementation would all stay the same so I don't particularly see the benefit. We'd have to start adding type checks to raise in the event that we're given `bytes`.

Reply:
> why

Because it's much simpler to reason about, and bytes should never be present in typical Python use. From the `os` docs:

> In Python, file names, command line arguments, and environment variables are represented using the string type. On some systems, decoding these strings to and from bytes is necessary before passing them to the operating system. Python uses the [filesystem encoding and error handler](https://docs.python.org/3/glossary.html#term-filesystem-encoding-and-error-handler) to perform this conversion (see [`sys.getfilesystemencoding()`](https://docs.python.org/3/library/sys.html#sys.getfilesystemencoding)).

So `sys.argv` and `sys.environ` will already be strings, and Python will handle if it needs to convert to bytes internally.